### PR TITLE
fix: [CDS-73357]: fix cleanup script to kill nested processes

### PR DESCRIPTION
**Pull request by hingerabhinav**
## Describe your changes
1. Clean child processes first, then parent
2. Get child processes as an array instead of string
3. Added logging for the cleaned up processes and parent, children for each iteration

This also checks if a process is alive or not before killing it. 

```
2023-07-18 10:44:53,087 [Thread-53] INFO  io.harness.shell.ScriptProcessExecutor - Parent process id: 50715  
2023-07-18 10:44:53,226 [Thread-53] INFO  io.harness.shell.ScriptProcessExecutor - Number of children processes: 2  
2023-07-18 10:44:53,227 [Thread-53] INFO  io.harness.shell.ScriptProcessExecutor - Children Process List Print:   
2023-07-18 10:44:53,227 [Thread-53] INFO  io.harness.shell.ScriptProcessExecutor - 50715 50716   
2023-07-18 10:44:53,227 [Thread-53] INFO  io.harness.shell.ScriptProcessExecutor - Processing Child Number: 1 with PID: 50716 and PPID: 50715  
2023-07-18 10:44:53,227 [Thread-53] INFO  io.harness.shell.ScriptProcessExecutor - Parent process id: 50716  
2023-07-18 10:44:53,333 [Thread-53] INFO  io.harness.shell.ScriptProcessExecutor - Number of children processes: 1  
2023-07-18 10:44:53,333 [Thread-53] INFO  io.harness.shell.ScriptProcessExecutor - Children Process List Print:   
2023-07-18 10:44:53,333 [Thread-53] INFO  io.harness.shell.ScriptProcessExecutor - 50716   
2023-07-18 10:44:53,335 [Thread-53] INFO  io.harness.shell.ScriptProcessExecutor - Termination successful for PID: 50716  
2023-07-18 10:44:53,338 [Thread-53] INFO  io.harness.shell.ScriptProcessExecutor - Process with PID 50715 is not running. Skipping termination.  
2023-07-18 10:44:53,340 [task-exec-4] INFO  io.harness.shell.ChildProcessStopper - Kill child processes command exited with: 0 [accountId=kmpySmUISimoRrJL6NL73w, taskId=Su5h_Cc0Txit7TPi3CuuAg] 

```

Tested 
1. Ansible processes 
2. process created by Sleep
3. Init processes. Since their parent is 1, we cant clean them. 
4. Processes created by C binary executables 

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/50217)

### CODE REVIEW

File: 960-api-services/src/main/java/io/harness/shell/ChildProcessStopper.java [hunk 1]
```diff
@@ -38,13 +38,15 @@ public void stop(Process process) {
...
-            + "  kill -9 ${children[0]}\n"
+            + "  local -a children=()\n"
+            + "  read -ra children <<< \"$(ps -ef | grep -v grep | grep $1 | awk '{print $2}' | tr '\\n' ' ')\"\n"
             + "\n"
             + "  for (( c=1; c<${#children[@]} ; c++ ))\n"
             + "  do\n"
             + "    list_descendants ${children[c]}\n"
             + "  done\n"
+            + "\n"
+            + "  kill -9 ${children[0]}\n"
```
Comment (acasian-harness) [inline 960-api-services/src/main/java/io/harness/shell/ChildProcessStopper.java:49]: shall we validate after we kill if there are no child processes hanging?

Comment (hingerabhinav) [reply 960-api-services/src/main/java/io/harness/shell/ChildProcessStopper.java]: @acasian-harness How do I check that? I have added the check for each process that I am cleaning up. Please check my change.

**Comment by hingerabhinav** [comment]
trigger uts

**Review by acasian-harness — APPROVED** [review]

**Comment by hingerabhinav** [comment]
trigger uts

File: 960-api-services/src/main/java/io/harness/shell/ChildProcessStopper.java [hunk 2]
```diff
@@ -38,20 +39,27 @@ public void stop(Process process) {
       try (FileOutputStream outputStream = new FileOutputStream(scriptFile)) {
         String command = "list_descendants ()\n"
             + "{\n"
-            + "  local children=$(ps -ef | grep -v grep | grep $1 | awk '{print $2}')\n"
-            + "  kill -9 ${children[0]}\n"
+            + "  local -a children=()\n"
+            + "  read -ra children <<< \"$(ps -ef | grep -v grep | grep $1 | awk '{print $2}' | tr '\\n' ' ')\"\n"
             + "\n"
```
Comment (yogesh-chauhan) [inline 960-api-services/src/main/java/io/harness/shell/ChildProcessStopper.java:44]: if there are no children, can we just verify if this kills the parent process or not

**Comment by hingerabhinav** [comment]
trigger uts

**Comment by hingerabhinav** [comment]
trigger uts

**Review by acasian-harness — DISMISSED** [review]

**Comment by hingerabhinav** [comment]
trigger uts

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 1 of 3 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50217)* status: 3 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained.

**Review by yogesh-chauhan — APPROVED** [review]

**Comment by hingerabhinav** [comment]
trigger review

**Comment by hingerabhinav** [comment]
trigger messagecheck

**Pull request merged**